Documentation/design-docs/dllmap.md

```diff
@@ -0,0 +1,246 @@
+
+## Expectations
+
+### .NET Core Dllmap
+
+.NET Core Dllmap’s purpose is to deliver a cross-platform support for p/invoke mechanism in .NET. With Dllmap user will be able to control interop methods by defining custom mapping between OS-specific dlls and methods.
+
+Dllmap will allow making changes in both library names and target method names (entrypoints). Changing entrypoint name will be optional (it will remain unchanged by default).
+
+Target platforms for this feature are: Windows, Linux and OS X.
+
+### Interaction with Dllmap
+
+#### Mono combability
+Dllmap should be easy to use. It’s possible to achieve this easily by keeping Mono-compatible style of XML mapping configuration file. It’s described [here](http://www.mono-project.com/docs/advanced/pinvoke/dllmap/). Thanks to combability with Mono, users will be able to migrate from Mono’s Dllmap to .NET Core cross-platform applications.
+
+#### Flexibility
+For users, who want to manage their cross-platform dll imports in their own way, dll-load specific events will be exposed. Users will be able to subscribe to these events and implement any loading policies. Details are described in the Design section.
+
+### Usage example (XML configuration)
+
+Let’s say there is a customer developing a .NET application with p/invokes on Windows. The customer wants the application to run on both Windows and Linux without introducing any changes in the code.
+
+The application calls a function GetCurrentProcessId from an OS-specific library kernel32.dll.
+```c++
+[DllImport("kernel32.dll")]
+static  extern  uint  GetCurrentProcessId();
+```
+
+To make it work on Linux, that does not have kernel32.dll, user must define a mapping of dll. There is no GetCurrentProcessId function in any corresponding Linux-specific library, so entrypoint name mapping must be defined too. To achieve this, the user puts an XML configuration file in the main application directory. The file looks like this:
+```c++
+<configuration>
+	<dllmap  dll="kernel32.dll">
+		<dllentry  dll="libc.so.6" name="GetCurrentProcessId" target="getpid" />
+	</dllmap>
+</configuration>
+```
+
+With this file, all GetCurrentProcessId calls get automatically mapped to getpid calls on runtime and the end user of the application can’t see any difference in application’s behavior. Running the application cross-platform does not require any OS-specific changes in the code. All the mapping is defined in advance in the external configuration file.
+
+This is a very basic scenario and it can be extended to different operating systems, libraries and entrypoints and to subscribing to dll specific events.
+
+## Design
+### XML configuration file
+For a basic case, mapping must be defined in an XML configuration file, that must be placed in the main application folder. It must be named application_name.xml where application_name is the name of the application.
```

Comment:
Mono also looks for AssemblyName.xml next to each assembly. The Mono applications that take advantage of this often have many .xml files - one file for each assembly that requires the remapping.

Reply:
True, thank you for pointing this out. We will do the same, I'll update the doc.

Reply:
We should consider support both(global and per assembly), similar to mono dllmap.